### PR TITLE
[feature](Nereids): eliminate groupby

### PR DESCRIPTION
**Pull request by jackwener**
## Proposed changes

eliminate groupby when unique constraint exist

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Override
+    public Rule build() {
+        return logicalAggregate()
+                .when(agg -> agg.getGroupByExpressions().stream().allMatch(expr -> expr instanceof Slot))
+                .then(agg -> {
+                    Set<Slot> groupby = agg.getGroupByExpressions().stream().map(e -> (Slot) e)
+                            .collect(Collectors.toSet());
+                    boolean unique = agg.getLogicalProperties().getFunctionalDependencies().isUnique(groupby);
+                    if (!unique) {
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java:52]: IsUniqueAndNotNull

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                    }
+
+                    List<NamedExpression> newOutput = agg.getOutputExpressions().stream().map(ne -> {
+                        if (ne instanceof Alias && ne.child(0) instanceof AggregateFunction) {
+                            AggregateFunction f = (AggregateFunction) ne.child(0);
+                            if (f instanceof Sum || f instanceof Min || f instanceof Max) {
+                                return new Alias(ne.getExprId(), f.child(0), ne.getName());
+                            } else if (f instanceof Count) {
+                                return (NamedExpression) ne.withChildren(
+                                        new If(new IsNull(f.child(0)), Literal.of(0), Literal.of(1)));
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java:69]: It's wrong for empty children

Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java:69]: It is right.

such as  for empty children.

```sql
create table test (id int, age int);

select min(id) from test group by age;
select id from test;
```

They are same;

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

File: conf/fe.conf [hunk 3]
```diff
@@ -69,6 +69,7 @@ arrow_flight_sql_port = -1
 # If no ip match this rule, will choose one randomly.
 # use CIDR format, e.g. 10.10.10.0/24 or IP format, e.g. 10.10.10.1
 # Default value is empty.
+priority_networks = 127.0.0.1/24
```
Comment (xzj7019) [inline conf/fe.conf:72]: remove this.

File: conf/be.conf [hunk 4]
```diff
@@ -55,7 +55,7 @@ ssl_private_key_path = "$DORIS_HOME/conf/key.pem"
 # If no ip match this rule, will choose one randomly.
 # use CIDR format, e.g. 10.10.10.0/24 or IP format, e.g. 10.10.10.1
 # Default value is empty.
-# priority_networks = 10.10.10.0/24;192.168.0.0/16
+priority_networks = 127.0.0.1/24
```
Comment (xzj7019) [inline conf/be.conf:58]: remove this.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/FunctionalDependencies.java [hunk 5]
```diff
@@ -220,7 +220,8 @@ public void replace(Map<Slot, Slot> replaceMap) {
                     .map(s -> replaceMap.getOrDefault(s, s))
                     .collect(Collectors.toSet());
             slotSets = slotSets.stream()
-                    .map(set -> set.stream().map(replaceMap::get).collect(ImmutableSet.toImmutableSet()))
+                    .map(set -> set.stream().map(s -> replaceMap.getOrDefault(s, s))
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/FunctionalDependencies.java:223]: what is it for ?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+public class EliminateGroupBy extends OneRewriteRuleFactory {
+
+    @Override
+    public Rule build() {
+        return logicalAggregate()
+                .when(agg -> agg.getGroupByExpressions().stream().allMatch(expr -> expr instanceof Slot))
+                .then(agg -> {
+                    Set<Slot> groupby = agg.getGroupByExpressions().stream().map(e -> (Slot) e)
+                            .collect(Collectors.toSet());
+                    boolean unique = agg.child().getLogicalProperties().getFunctionalDependencies()
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java:51]: pls also consider the uk model also, besides fd injection.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/FunctionalDependencies.java (hunk 5 above)
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/FunctionalDependencies.java:223]: bugfix

**Comment by jackwener** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java (hunk 6 above)
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java:51]: cc @XieJiann

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run External Regression

**Comment by jackwener** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java (hunk 6 above)
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupBy.java]: yes

**Review by xzj7019 — APPROVED** [review]

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

**Review by starocean999 — APPROVED** [review]

**Pull request merged**